improver/utilities/common_input_handle.py

```diff
@@ -0,0 +1,62 @@
+# (C) Crown copyright, Met Office. All rights reserved.
+#
+# This file is part of IMPROVER and is released under a BSD 3-Clause license.
+# See LICENSE in the root of the repository for full licensing details.
+from typing import Union
+
+from iris.cube import Cube, CubeList
+
+from improver.utilities.flatten import flatten
+
+
+def as_cubelist(*cubes: Union[Cube, CubeList]):
+    """
+    Standardise input handling of cube/cubelist arguments.
+
+    The role of this function is to flatten the provided inputs and thereby
+    return a single CubeList object.
+
+    Args:
+        cubes:
+            Input data provided in the form of one or more cubes or cubelists (or mixture thereof).
+            Any iterable is supported.
+
+    Returns:
+        CubeList:
+            A CubeList containing all the cubes provided as input.
+    """
+    cubes = CubeList(flatten(cubes))
+    # Remove CubeList verification for iris >=3.3.0
+    for cube in cubes:
+        if not hasattr(cube, "add_aux_coord"):
```

Comment:
In recent versions of iris, `hasattr` is the approach taken within the CubeList cube check built-in:
https://github.com/SciTools/iris/blob/b15a1f61b7d48ca9c7af464c5a5f112ac7366bc7/lib/iris/cube.py#L154-L158

When we upgrade to newer iris, we will be subject to iris CubeList built-in behaviour anyway.  I have preferenced not deviating from later iris version behaviour to avoid hiding potential issues that may arise later from the upgrade.

As an aside, the 'pythonic' approach is indeed to generally avoid `isinstance` calls where possible.  The idea is that if something has the interface you need, then it doesn't matter what the object actually is (e.g. code handling numpy/dask arrays).
However, in this case, I actually very much agree with you that an `isinstance` call is preferable in iris CubeList cube checking :(
(cubelist behaviour is not limited to an add_aux_coord interface of the object and so is not indication enough of the validity of the objects it contains)

Reply:
OK, I wouldn't have expected that checking for this attribute would be how iris would identify cubes or cubelists internally. As you say, we may as well adhere to this form of checking given that this is what is done in the newer versions of Iris. I hadn't appreciated that checking for an attribute in this way would be considered more "pythonic" but I understand that just checking for the existence of a particular attribute, if that is your only concern, would be more flexible. I'm not sure that raising this as an iris issue will necessarily be worthwhile, so I'm happy for you to leave this.